src/components/tooltip/GCMarker.js

```diff
@@ -423,80 +430,104 @@ function _treeInsert(
   let node = tree.get(component);
   if (!node) {
     // Make a new node and grow the tree in this direction.
-    node = { branches: new Map() };
+    const value = { name: '(temp)', time: 0 };
+    node = { branches: new Map(), leaf: false, value };
 
     tree.set(component, node);
   }
 
   if (path.length > 0) {
     // There are more path components.  This node should be a branch if it
     // isn't one already.
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== '(temp)') {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
-function _treeGetLeaves(
-  tree: Map<string, PhaseTreeNode>
-): Array<PhaseTimeTuple> {
-  const leaves = [];
+function _treeMapNodes<T>(
```

Comment:
I find this name a bit confusing, also due do the fact we use this in 2 different ways: one where we loop and actually mutate the tree, and one where we filter and flatten the tree and use the returned array.

Maybe we can keep this function as it is, with a name that better reflect what it does: what about `_flattenTreeWithFiltering` ?
But _also_ have another function that would only loop over the tree, without returning anything, with maybe a name `_forEachTreeNode`. Of course it could just reuse `_flattenTreeWithFiltering` inside, with a nice comment explaining a bit of all this. Or reimplement it, because after all, this is super simple too.
Then at the location where we use them, it would be clearer, I believe.

The term `map` itself is confusing, because in JS a `map` operation wouldn't return a different structure (at least using the semantics I know in JS).

Reply:
(I ended up creating a `_forEachTreeNode` and dropping `_treeMapNodes`/`_flattenTreeWithFiltering`.)